### ISSUE TITLE
php7.2 - libsodium-dev version differs between Debian and Ubuntu

### ISSUE THREAD

**Issue by peios** [oerdnj/deb.sury.org#770]
## General Information

__Package:__ php7.2
__Tested Under:__ Debian 9.3 & Ubuntu 14.04
__Web Server Environment:__ nginx (the version should not be relevant for this issue)

## TL;DR
Is there any way for you to ensure that both the Debian and the Ubuntu builds for PHP7.X employ the same libsodium-dev version in the future?

## Summary
The libsodium package defines different constants to control both MEMLIMIT and OPSLIMIT for hashing operations. These constants are (kind of) encoded into the resulting hash string and are employed to be able to keep track of the parameters employed for the computation of the specific hash.

Obviously, changing this constants between two computations causes a totally different result, as it is expected and desired with hashing operations.

It appears that a **huge** breaking change was introduced between version *1.0.13* and *1.0.15* of the *libsodium* package, which makes the re-computation impossible if two systems are found to employ these different versions. This happened today with Debian 9.3 and Ubuntu 14.04. PHP 7.2 was installed from your PPAs.

## Details
For higher consistency, the PHP library simply reads libsodium constants from the underlying system library ([an example of what I mean](http://git.php.net/?p=php-src.git;a=blob;f=ext/sodium/libsodium.c;h=ae93e6dc58622c40d11b983ddeab52dd5fb7368f;hb=5cfa26c18189ae5e0ae8bb1eac5dd0e213a2bb3e#l390)). This ensures that, when the underlying implementation changes, the library promptly reacts to this change.

Unfortunately, [this unlucky commit](https://github.com/jedisct1/libsodium/commit/c8425e63861e6f855408abd9130df3aa14b66887#diff-08ad145127ef7fd793170380faf9ceb4L49) introduced a change in the constants I cited above.

Now, to make it shorter:
 - I register an account on host A (Ubuntu, sodium 1.0.15) and have my password hashed and stored in a database. The hashing is performed using the **INTERACTIVE** settings for both OPSLIMIT and MEMLIMIT to make it simple.
 - I go for the login and end up on host B (Debian, sodium 1.0.14)
 - I provide correct username and password.
 - I watch `sodium_crypto_pwhash_str()` return `false` in front of me. Why?

## Why?

The endurance constants changed between the two versions. What in 1.0.15 means `OPSLIMIT_SENSITIVE` (a.k.a.: `(int) 2`), in 1.0.14 is **invalid** since smaller than `OPSLIMIT_MIN` (a.k.a.: `(int) 3`).

## *Why are you opening this issue here?*

What I feel sorry about asking is: is there any possibility for you to make sure that future builds of the php7.2 package will be built with the same `libsodium-dev` version for both Debian and Ubuntu? This would fix the incongruence, introduced by a rather careless breaking change between patch releases (and I add: -__-).

Of course, no, `apt install libsodium-dev=specific-version` did not solve the issue, as the dependency seems to get automatically shipped with the extension, which is in turn now automatically shipped with the php7.2 package.

## *Closing this as {your-reason-here}*

No problem, and I feel sorry if this is not the right place for this issue! Just please, provide a reason so I won't repeat the mistake :)

**Comment by oerdnj** [comment]
TL;DR - it's a problem in libsodium, but sure, I'll try to keep them aligned.

I just updated them both to 1.0.16-2 (synced with Debian), and PHP 7.2 will get rebuild when 7.2.1 has been released (this should happen this week, so I am closing the bug).

**Comment by peios** [comment]
Thank you :) should you need further details, please reopen the issue, I'll be on it!On Jan 3, 2018 21:29, Ondřej Surý <notifications@github.com> wrote:TL;DR - it's a problem in libsodium, but sure, I'll try to keep them aligned.
I just updated them both to 1.0.16-2 (synced with Debian), and PHP 7.2 will get rebuild when 7.2.1 has been released (this should happen this week, so I am closing the bug).
